cadcutils/cadcutils/util/utils.py

```diff
@@ -172,43 +173,155 @@ def parse_resource_id(resource_id):
+        else:
+            super(_CustomArgParser, self).__init__(**kwargs)
+
+    def add_subparsers(self, **kwargs):
+        if not self.subparsers:
+            raise RuntimeError('Parser created to run without subparsers')
+        self._subparsers_added = True
+        return _AugmentAction(self.common_parser,
+                        super(_CustomArgParser, self).add_subparsers(**kwargs))
+
+    def parse_args(self, args=None, namespace=None):
+         if self.subparsers and not self._subparsers_added:
+              raise RuntimeError('No subparsers added. Change the parsers flag?')
+         return super(_CustomArgParser, self).parse_args(args=args, namespace=namespace)
+
+
+
+def get_base_parser(subparsers=True, version=None, usecert=True, default_resource_id=None):
+    """
+    An ArgumentParser with some common things most CADC clients will want. There are two
+    modes to use this parser: with or without subparsers. With supbarsers (subparsers=True),
+    separate subparsers are created for each subcommand and the common CADC options show
+    up in each subcommand (and not on the parent parser). Without subparsers (subparsers=False)
+    the common options are automatically added to the base parser.
+
+    :param subparsers: True if the parser will use subparsers (subcommands) otherwise False
     :param version: A version number if desired.
-    :param usecert: If True add '--certfile' argument.
+    :param usecert: If True add '--cert' argument.
     :param default_resource_id: default resource identifier to use
     :return: An ArgumentParser instance.
     """
-    parser = ArgumentParser(add_help=False)
+    cparser = ArgumentParser(add_help=False, formatter_class=SingleMetavarHelpFormatter)
+
+    if version is not None:
+        cparser.add_argument('-V', '--version', action='version', version=version)
+    auth_group = cparser.add_mutually_exclusive_group()
     if usecert:
-        parser.add_argument('--certfile', type=str,
-                            help="location of your CADC certificate "
-                            + "file (default: $HOME/.ssl/cadcproxy.pem, " +
-                            "otherwise uses $HOME/.netrc for name/password)",
-                            default=os.path.join(os.getenv("HOME", "."),
-                                                 ".ssl/cadcproxy.pem"))
-    parser.add_argument('--anonymous', action="store_true",
-                        help='Force anonymous connection')
-    parser.add_argument('--host', help="Base hostname for services - used mainly for testing " +
-                                       "(default: www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca)",
+        auth_group.add_argument('--cert', type=str,
+                            help='location of your X509 certificate to use for authentication ' +
+                            '(unencrypted, in PEM format)')
+    auth_group.add_argument('-n', action='store_true', help='Use .netrc in $HOME for authentication')
```

Comment:
The base parse gives me the impression that commands will all need to specify -n (which normally would be 'no' or 'not' ) if they want the .netrc to be used?  And also would need to give --cert if they want the certificate file to be checked?   I'm strongly in favour of looking for those settings in their default locations but allowing users to override the defaults or turn off using those values as needed...  so, --anonymous is good as that turns off all authentication, then maybe --ignore-netrc  would be the other one to add, if we want to force ignoring the .netrc file and --ignore-cert to not look at the certificate in the default location and then --netrc-file XXX --cert-file 'xxxx' to set non-default values.  (which then implies --ignore-cert or --ignore-netrc)